EIPS/eip-3540.md

```diff
@@ -0,0 +1,218 @@
+- Or completely remove the need for `JUMPDEST` instructions.
+- This helps with deprecating EVM instructions and/or features.
+- The biggest disadvantage is that deploy-time validation of EOF code must be enabled in two hard-forks.
+
+The alternative is to have execution time validation for EOF. This is performed every single time a contract is executed, however clients may be able to cache validation results. This approach has the following properties:
+- Because the validation is consensus-level execution step, it means the execution always requires the entire code. This makes code merkleization impractical.
+- The main advantage is it can be enabled via a single hard-fork.
+- A second advantage is better backwards compatibility: data contracts starting with the `0xEF` byte or the *EOF prefix* can be deployed. This is a dubious benefit however.
+
+### Initcode vs code
+
+After HF1 and before HF2 the `FORMAT` first byte check only applies to _code_. Applying the rule also to _initcode_ would have the same effects because is not distinguishable from executing _initcode_: if it starts with `FORMAT` execution would exceptionally abort. Yet, we decided against introducing an explicit check for _initcode_ for the simplicity of the HF1 spec.
+
+### Contract creation restrictions
+
+The [Changes to contact creation semantics](#changes-to-contract-creation-semantics) section defines minimal set of restrictions related to the contract creation: if _initcode_ or _code_ has the EOF1 container prefix it must be validated. This adds two validation steps in the contract creation, any of it failing will result in contract creation failure.
+
+Because _initcode_ and _code_ is evaluated for EOF1 independently, number of interesting combinations are allowed:
+- Create transaction with EOF1 _initcode_ can deploy legacy contract,
+- EOF1 contract can execute `CREATE` instruction with legacy _initcode_ to create new legacy contract,
+- Legacy contract can execute `CREATE` instruction with EOF1 _initcode_ to create new EOF1 contract,
+- Legacy contract can execute `CREATE` instruction with EOF1 _initcode_ to create new legacy contract,
+- etc.
+
+To limit the number of exotic bytecode version combinations, additional restrictions are considered, but current not part of the specification:
+
+1. The EOF version of _initcode_ must much the version of _code_.
+2. An EOF1 contract must not create legacy contracts.
+
+Finally, create transaction must be allowed to contain legacy _initcode_ and deploy legacy _code_ because otherwise there is no transition period allowing upgrading transaction signing tools. Deprecating such transactions may be considered in future.
+
+### The FORMAT byte
+
+The `0xEF` byte was chosen because it resembles **E**xecutable **F**ormat. It has a long history of being proposed for this use case, starting with [this](https://github.com/ethereum/EIPs/issues/154).
```

Comment:
```suggestion
 The `0xEF` byte was chosen because it resembles **E**xecutable **F**ormat.
```
Don't include external links.  Also, having a long history of being proposed for a case doesn't really mean anything so I recommend just leaving that bit out entirely.